### PR TITLE
librbd: fix inconsistency between AioCompletion is_complete() and wait_for_complete()

### PR DESCRIPTION
**Pull request by idryomov**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by idryomov** [comment]
Strictly speaking this is a bug fix, but `rbd_aio_is_complete()`/`RBD::AioCompletion::is_complete()` is unlikely to be used in the wild, so it's probably not worth backporting.

**Comment by idryomov** [comment]
> Agent went offline during the build

jenkins test make check

**Comment by idryomov** [comment]
jenkins test make check

**Comment by idryomov** [comment]
jenkins test make check

**Comment by idryomov** [comment]
jenkins test make check

**Comment by idryomov** [comment]
I had to retrigger `make check` a few times but I haven't seen any spurious failures in `ImagePollIO` or other RBD tests.

**Review by ajarr — COMMENTED** [review]
Commit message says, "... the only place that distinguishes between AIO_STATE_CALLBACK and AIO_STATE_COMPLETE is is_complete() ... "
`is_complete()` here should be `wait_for_complete()`?  Only `wait_for_complete()` checks for AIO_STATE_COMPLETE.

**Comment by idryomov** [comment]
> Commit message says, "... the only place that distinguishes between AIO_STATE_CALLBACK and AIO_STATE_COMPLETE is is_complete() ... "
> `is_complete()` here should be `wait_for_complete()`?

No, I meant to say `is_complete()`.

> Only `wait_for_complete()` checks for AIO_STATE_COMPLETE.

I used the word distinguish in the "who cares about the fact that apart from `AIO_STATE_COMPLETE` there is also `AIO_STATE_CALLBACK`" sense.  Or, alternatively, "whose behavior would change if one removes `AIO_STATE_CALLBACK` without changing anything else".  In this case, it's `is_complete()` because it checks for `!= AIO_STATE_PENDING`, which means that its return value changes (or rather used to change before this PR) when the completion moves to `AIO_STATE_CALLBACK`, before the move to `AIO_STATE_COMPLETE`.

I'm open to suggestions on how to word it better.

**Comment by ajarr** [comment]
Maybe something like the following:

Currently, there is inconsistency between public facing APIs `is_complete()` and `wait_for_complete()`. The ` is_complete()` API returns true for states AIO_STATE_CALLBACK and AIO_STATE_COMPLETE, while `wait_for_complete()`  blocks on
AIO_STATE_CALLBACK and returns on AIO_STATE_COMPLETE.

Since commit 002afa0 ("librbd: avoid using lock within AIO completion where possible"), the transient callback state AIO_STATE_CALLBACK is not essential. If we set state AIO_STATE_COMPLETE at the top of mark_complete_and_notify() (i.e. before event socket notification), AIO_STATE_CALLBACK can be eliminated and the inconsistency between the public APIs goes away.

**Comment by ajarr** [comment]
I realize that my suggestion skips your explanation
> I used the word distinguish in the "who cares about the fact that apart from AIO_STATE_COMPLETE there is also AIO_STATE_CALLBACK" sense. Or, alternatively, "whose behavior would change if one removes AIO_STATE_CALLBACK without changing anything else". In this case, it's is_complete() because it checks for != AIO_STATE_PENDING, which means that its return value changes (or rather used to change before this PR) when the completion moves to AIO_STATE_CALLBACK, before the move to AIO_STATE_COMPLETE.

I'm not sure if it's worth going into such detail. I found the `the only place that distinguishes between AIO_STATE_CALLBACK and AIO_STATE_COMPLETE is is_complete() `  hard to follow.  I leave it up to you.

**Review by nbalacha — APPROVED** [review]

**Comment by idryomov** [comment]
@ajarr I changed the description to

> After commit 002afa0fe375 ("librbd: avoid using lock within AIO
   completion where possible"), the only method whose behavior would
   change if AIO_STATE_CALLBACK is removed is is_complete() and it
   actually needs fixing anyway: because of state != AIO_STATE_PENDING
   test, is_complete() returns true both for AIO_STATE_CALLBACK and
   AIO_STATE_COMPLETE, while wait_for_complete() still blocks on
   AIO_STATE_CALLBACK and returns only on AIO_STATE_COMPLETE.

and also added
```
@@ -237,7 +237,7 @@ void AioCompletion::complete_request(ssize_t r)
 
 bool AioCompletion::is_complete() {
   tracepoint(librbd, aio_is_complete_enter, this);
-  bool done = (this->state != AIO_STATE_PENDING);
+  bool done = (this->state == AIO_STATE_COMPLETE);
   tracepoint(librbd, aio_is_complete_exit, done);
   return done;
 }
```
hunk to make both methods check for `AIO_STATE_COMPLETE` explicitly (with only two states left, `!= AIO_STATE_PENDING` and `== AIO_STATE_COMPLETE` tests are equivalent).

**Review by ajarr — APPROVED** [review]

**Comment by ajarr** [comment]
> @ajarr I changed the description to
> 
> > After commit [002afa0](https://github.com/ceph/ceph/commit/002afa0fe37544ab4eb15f973e00068e1240c0ef) ("librbd: avoid using lock within AIO
> > completion where possible"), the only method whose behavior would
> > change if AIO_STATE_CALLBACK is removed is is_complete() and it
> > actually needs fixing anyway: because of state != AIO_STATE_PENDING
> > test, is_complete() returns true both for AIO_STATE_CALLBACK and
> > AIO_STATE_COMPLETE, while wait_for_complete() still blocks on
> > AIO_STATE_CALLBACK and returns only on AIO_STATE_COMPLETE.  These
> > methods back public APIs, so this inconsistency is exposed to users.
> 
> and also added
> 
> ```
> @@ -237,7 +237,7 @@ void AioCompletion::complete_request(ssize_t r)
>  
>  bool AioCompletion::is_complete() {
>    tracepoint(librbd, aio_is_complete_enter, this);
> -  bool done = (this->state != AIO_STATE_PENDING);
> +  bool done = (this->state == AIO_STATE_COMPLETE);
>    tracepoint(librbd, aio_is_complete_exit, done);
>    return done;
>  }
> ```
> 
> hunk to make both methods check for `AIO_STATE_COMPLETE` explicitly (with only two states left, `!= AIO_STATE_PENDING` and `== AIO_STATE_COMPLETE` tests are equivalent). Let me know if it reads better!

These changes make it clearer. Thanks!

**Comment by idryomov** [comment]
jenkins test api

**Comment by idryomov** [comment]
No related failures:

https://pulpito.ceph.com/dis-2024-07-18_22:11:30-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2024-07-19_12:06:54-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2024-07-19_20:40:00-rbd-wip-dis-testing-distro-default-smithi/

(Jobs where `TestLibRBD.TestEncryptionLUKS1` triggers an assert
```
2024-07-19T12:48:08.399 INFO:tasks.ceph.osd.3.smithi110.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-5126-g01e1aa05/rpm/el9/BUILD/ceph-19.0.0-5126-g01e1aa05/src/osd/ECCommon.cc: In function 'static void ECCommon::ReadPipeline::get_min_want_to_read_shards(uint64_t, uint64_t, const ECUtil::stripe_info_t&, const std::vector<int>&, std::set<int>*)' thread 7f9b60945640 time 2024-07-19T12:48:08.398035+0000
2024-07-19T12:48:08.399 INFO:tasks.ceph.osd.3.smithi110.stderr:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos9/DIST/centos9/MACHINE_SIZE/gigantic/release/19.0.0-5126-g01e1aa05/rpm/el9/BUILD/ceph-19.0.0-5126-g01e1aa05/src/osd/ECCommon.cc: 336: FAILED ceph_assert(want_to_read->size() == sinfo.get_data_chunk_count())
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: ceph version 19.0.0-5126-g01e1aa05 (01e1aa052c15e2a077a20de402d3ae763de4a30d) squid (dev)
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x11e) [0x56123c9a347e]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 2: ceph-osd(+0x3fd63a) [0x56123c9a363a]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 3: ceph-osd(+0x3a4a88) [0x56123c94aa88]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 4: (ECCommon::ReadPipeline::get_min_want_to_read_shards(unsigned long, unsigned long, std::set<int, std::less<int>, std::allocator<int> >*)+0x56) [0x56123cc9b886]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 5: (ECCommon::ReadPipeline::objects_read_and_reconstruct(std::map<hobject_t, std::__cxx11::list<ECCommon::ec_align_t, std::allocator<ECCommon::ec_align_t> >, std::less<hobject_t>, std::allocator<std::pair<hobject_t const, std::__cxx11::list<ECCommon::ec_align_t, std::allocator<ECCommon::ec_align_t> > > > > const&, bool, std::unique_ptr<GenContext<std::map<hobject_t, ECCommon::ec_extent_t, std::less<hobject_t>, std::allocator<std::pair<hobject_t const, ECCommon::ec_extent_t> > >&&>, std::default_delete<GenContext<std::map<hobject_t, ECCommon::ec_extent_t, std::less<hobject_t>, std::allocator<std::pair<hobject_t const, ECCommon::ec_extent_t> > >&&> > >&&)+0x8f7) [0x56123cca48a7]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 6: (ECBackend::objects_read_async(hobject_t const&, std::__cxx11::list<std::pair<ECCommon::ec_align_t, std::pair<ceph::buffer::v15_2_0::list*, Context*> >, std::allocator<std::pair<ECCommon::ec_align_t, std::pair<ceph::buffer::v15_2_0::list*, Context*> > > > const&, Context*, bool)+0x5a0) [0x56123ceb8240]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 7: (PrimaryLogPG::OpContext::start_async_reads(PrimaryLogPG*)+0x179) [0x56123cbda069]
2024-07-19T12:48:08.401 INFO:tasks.ceph.osd.3.smithi110.stderr: 8: (PrimaryLogPG::execute_ctx(PrimaryLogPG::OpContext*)+0x474) [0x56123cc0adc4]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 9: (PrimaryLogPG::do_op(boost::intrusive_ptr<OpRequest>&)+0x342b) [0x56123cc036ab]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 10: (OSD::dequeue_op(boost::intrusive_ptr<PG>, boost::intrusive_ptr<OpRequest>, ThreadPool::TPHandle&)+0x197) [0x56123cb41697]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 11: (ceph::osd::scheduler::PGOpItem::run(OSD*, OSDShard*, boost::intrusive_ptr<PG>&, ThreadPool::TPHandle&)+0x69) [0x56123cd864d9]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 12: (OSD::ShardedOpWQ::_process(unsigned int, ceph::heartbeat_handle_d*)+0xd07) [0x56123cb5ade7]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 13: (ShardedThreadPool::shardedthreadpool_worker(unsigned int)+0x2aa) [0x56123d05ef6a]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 14: ceph-osd(+0xab9514) [0x56123d05f514]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 15: /lib64/libc.so.6(+0x89c02) [0x7f9b83a89c02]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr: 16: /lib64/libc.so.6(+0x10ec40) [0x7f9b83b0ec40]
2024-07-19T12:48:08.402 INFO:tasks.ceph.osd.3.smithi110.stderr:*** Caught signal (Aborted) **
```
in the OSD were excluded from last the rerun.  This issue is being looked at by the RADOS team.)
